### PR TITLE
US122429 Quiz password ui

### PR DESCRIPTION
**Pull request by ronan-f**
https://rally1.rallydev.com/#/29180338367d/iterationstatus?detail=%2Fuserstory%2F456674904728

### CODE REVIEW

File: components/d2l-activity-editor/d2l-activity-quiz-editor/lang/en.js [hunk 1]
```diff
@@ -7,8 +7,9 @@ export default {
 	"hdrTimingAndDisplay": "Timing & Display", // timing/display accordion header
 	"hdrAttemptsAndCompletion": "Attempts & Completion", //attempts/completion accordion header
 	"displayTools": "Display", // Title for display tools
-	"hintsToolDescription": "Allow hints", // description for allow hints
-	"hintsAllowedSummary": "Hints allowed", // summary text when hints are enabled
+	"hintsToolDescription": "Allow hints", // decription for hints tool
+	"hintsAllowedSummary": "Hints allowed", // summary text when hints are enabled,
+	"passwordLabel": "Password", // Lable for the quiz password text input field
```
Comment (mostlyfabulous) [inline components/d2l-activity-editor/d2l-activity-quiz-editor/lang/en.js:12]: Typo nit on code comment `// Lable`

File: components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js [hunk 2]
```diff
@@ -1,6 +1,4 @@
 import '../d2l-activity-accordion-collapse.js';
-import '../d2l-activity-availability-dates-editor.js';
-import '../d2l-activity-availability-dates-summary.js';
```
Comment (mostlyfabulous) [inline components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js:3]: Where did the summary go? I may have missed it.

**Comment by mostlyfabulous** [comment]
Not sure how many tests we should add for this but I would suggest:

1. Expect failure when setting password and `canEditPassword` is `false`
2. Expect failure when setting password to a password `> 20` chars and `canEditPassword` is `true`

Probably don't need more as I'm hoping the `<d2l-input-text>` component has checks on empty string, weird characters and other edge cases.

**Comment by ronan-f** [comment]
> Not sure how many tests we should add for this but I would suggest:
> 
> 1. Expect failure when setting password and `canEditPassword` is `false`
> 2. Expect failure when setting password to a password `> 20` chars and `canEditPassword` is `true`
> 
> Probably don't need more as I'm hoping the `<d2l-input-text>` component has checks on empty string, weird characters and other edge cases.

Yip good suggestions. To address them:

1. So the saving of a password happens in the `siren-sdk` repo and there's a check to ensure that a user has edit permissions at that point. Updating the mobx store does not perform such a check so I don't see that we can write a test for this. Also, the input component _should_ be disabled and prevent input if the user does not have the correct permissions.
2. There's no logic in any of the functions I've added that will restrict input to 20 chars. This is implemented in the `<d2l-input-text>` component which seems to have pretty [comprehensive test coverage](https://github.com/BrightspaceUI/core/blob/master/components/inputs/test/input-text.test.js#L215).

Would be curious to hear other thoughts around maybe adding some additional checks for these things in this repo.

File: components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js (hunk 2 above)
Comment (ronan-f) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js:3]: Yeah neither the dates editor or summary are rendered in this component so no need for these imports. Probably got left over from a previous change.

File: components/d2l-activity-editor/d2l-activity-quiz-editor/lang/en.js (hunk 1 above)
Comment (ronan-f) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/lang/en.js:12]: Good catch! Updated.

File: components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js (hunk 2 above)
Comment (mostlyfabulous) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-timing-and-display-editor.js:3]: Good to know

File: test/d2l-activity-editor/d2l-activity-quiz-editor/state/quiz.spec.js [hunk 3]
```diff
@@ -74,4 +76,12 @@ describe('Quiz', function() {
 		expect(quiz.name).to.equal('No Homework');
 	});
 
+	it('setPassword', async() => {
+		const quiz = new Quiz('http://quiz/1', 'token');
+		await quiz.fetch();
+		quiz.setPassword('super-secret-passord');
```
Comment (wongvincent) [inline test/d2l-activity-editor/d2l-activity-quiz-editor/state/quiz.spec.js:82]: nit. typo on `password`

File: components/d2l-activity-editor/d2l-activity-quiz-editor/state/quiz.js [hunk 4]
```diff
@@ -87,11 +94,13 @@ decorate(Quiz, {
 	canEditHints: observable,
 	canEditDisableRightClick: observable,
 	hintsToolEnabled: observable,
+	password: observable,
```
Comment (wongvincent) [inline components/d2l-activity-editor/d2l-activity-quiz-editor/state/quiz.js:97]:
```suggestion
	password: observable,
	canEditPassword: observable,
```

File: components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-availability-editor.js [hunk 5]
```diff
@@ -64,7 +67,7 @@ class ActivityQuizAvailabilityEditor extends AsyncContainerMixin(LocalizeActivit
 		return html`
 			<div class="d2l-editor">
 				<d2l-activity-availability-dates-editor
-					href="${this.href}"
+					href="${this.activityUsageHref}"
```
Comment (wongvincent) [inline components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-availability-editor.js:70]: Are these `activityUsageHref` changes supposed to be part of this PR?

Comment (ronan-f) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-availability-editor.js:70]: Yeah I believe so. `<d2l-activity-availability-dates-summary>` needs the `activityUsageHref` and `<d2l-activity-quiz-password-editor>` needs the specific quizzing href from what I could tell. Would love to know if there's a better way to do that?

Comment (ronan-f) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-availability-editor.js:70]: The dates-editor accesses `activity-store.js` as seen [here ](https://search.d2l.dev/xref/GitHub/BrightspaceHypermediaComponents/activities/components/d2l-activity-editor/d2l-activity-availability-dates-editor.js?r=bd2c687d#8) which to my understanding means that it would need an `activityUsageHref` whereas the password editor uses `quiz-store.js` which is expecting the specific quizzing  href.

Comment (wongvincent) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-availability-editor.js:70]: yeah, I see now that 
`d2l-activity-quiz-editor-secondary.js` is passing in the quiz href as `this.href` so I understand why this is changed now to use `this.activityUsageHref`.

**Review by wongvincent — APPROVED** [review]

File: components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-password-editor.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+			<d2l-input-text
+				?skeleton="${this.skeleton}"
+				id="quiz-password"
+				maxlength="20"
+				value="${entity.password}"
+				@input="${this._setPassword}"
+				label="${this.localize('passwordLabel')}"
+				?disabled="${!entity.canEditPassword}"
+				prevent-submit>
+			</d2l-input-text>
```
Comment (mdulat) [inline components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-password-editor.js:48]: Sorry I'm late to the game here, but should this input have `type="password"`? I think that just hides the input while the user types it in. Maybe a question for Patrick or Joseph.

Comment (ronan-f) [reply components/d2l-activity-editor/d2l-activity-quiz-editor/d2l-activity-quiz-password-editor.js:48]: Yeah when I first started this story I spoke with Joseph about password security (do we need to hash it/hide it etc) and my understanding was that it's not treated as a "secure" password but more like how a Zoom password would work. Looking at classic quizzing the password is not hidden so I suspect we should have the same behaviour. 

i will confirm this with Joseph though.
